agent/csec_metadata.h

```diff
@@ -0,0 +1,12 @@
+#include "php_agent.h"
+#include "util_hashmap.h"
+
+#define KEY_ENTITY_NAME "entity.name"
+#define KEY_ENTITY_TYPE "entity.type"
+#define KEY_ENTITY_GUID "entity.guid"
+#define KEY_HOSTNAME "hostname"
+#define KEY_AGENT_RUN_ID "agent.run.id"
+#define KEY_ACCOUNT_ID "account.id"
+#define KEY_LICENSE "license"
+#define KEY_PLICENSE "plicense"
+#define KEY_HIGH_SECURITY "high_security"
```

Comment:
Are all these values really required by csec-agent? Why? When does the csec-agent going to use these values?

Reply:
> Yes, the backend (Security Engine) uses these values to correctly map all the information for application, incidents, etc. csec-agent sends this information with the events.

I really doubt that backend will use `plicense` to map anything. Why is this particular value needed?

> If the value changes, as mentioned in the documentation for given method, same impact will be there. But we will not be missing any vulnerabilities due to this.

I do not understand your answer. If csec-agent is going to retrieve this information before `[newrelic_set_appname](https://docs.newrelic.com/docs/apm/agents/php-agent/php-agent-api/newrelic_set_appname/)` is called in PHP code, and send the 'event' to Security Engine after `[newrelic_set_appname](https://docs.newrelic.com/docs/apm/agents/php-agent/php-agent-api/newrelic_set_appname/)` is called, then the 'event' will not be correctly mapped. When does csec-agent retrieve the metadata and when does csec-agent generate the events sent to the backend? This is very important design consideration.

Reply:
We are logging some of these values, and that includes the license as well. Logging for license in particular has to be masked, and hence the reason for plicense.

At any point of time, when the apm agent is connected, entity.name is going to have a value along with entity.guid. When the newrelic_set_appname is called, csec-agent will be using that value on next reconnect cycle to the server. So, yes as soon as newrelic_set_appname is called, entity.name and entity.guid will not change for the csec-agent but on the next reconnect (this is planned at SE, the backend), the events will be mapped to new entity.name and entity.guid, this would reset the values for all the running and connected processes.

csec-agent tries to retrieve the metadata at the beginning of each request until it has data. It will wait for next reconnect cycle with server to see if anything has changed.

csec-agent generates the events based on the hooks and send the data immediately to the backend.

As has been detailed the intent of using newrelic_set_appname as early as possible in code, taking runtime values from APM agent would not affect the app's data at backend.

Reply:
> csec-agent tries to retrieve the metadata at the beginning of each request until it has data.

This is exactly why php-agent team has a problem with this. Not only there's potential for incorrect mapping of an event to an app, but also this depends on wether php-agent has been initialized before csec-agent, which we cannot see how it can be guaranteed. Why doesn't csec-agent retrieve the metadata only at the time it is needed, in the 'hook' you mention here:

> csec-agent generates the events based on the hooks and send the data immediately to the backend.

Reply:
csec-agent takes the connection information from the apm agent only. It does not generate any data until the apm agent has been connected. There is no way csec-agent can get entity.guid from any source other than APM agent.

Reply:
And without entity.guid, it does not connect to the backend